lxd/instance/drivers/driver_qemu.go

```diff
@@ -2589,6 +2589,158 @@ func (d *qemu) nvramPath() string {
 	return filepath.Join(d.Path(), "qemu.nvram")
 }
 
+// PyUEFIVars represents the JSON output format of the python-uefivars utility
+// URL: https://github.com/awslabs/python-uefivars
+type PyUEFIVars struct {
+	Version   uint32      `json:"version"`
+	Variables []PyUEFIVar `json:"variables"`
+}
+
+// PyUEFIVar represents a EFI variable entry
+type PyUEFIVar struct {
+	// EFI variable name
+	GUID string `json:"guid"`
+
+	// EFI variable name
+	Name string `json:"name"`
+
+	// EFI variable data (HEX-encoded)
+	Data string `json:"data"`
+
+	// EFI variable attributes
+	Attr uint32 `json:"attr"`
+
+	// EFI variable timestamp (HEX-encoded)
+	Timestamp string `json:"timestamp"`
+
+	// EFI variable digest (HEX-encoded)
+	Digest string `json:"digest"`
+}
+
+func callPyUEFIVars(stdin io.Reader, stdout io.Writer, efiVarsPath string) error {
+	var pyUEFIVarsCmd string
+	if stdin == nil {
+		pyUEFIVarsCmd = fmt.Sprintf("uefivars.py -o json -i edk2 -I %s", efiVarsPath)
+	} else {
+		pyUEFIVarsCmd = fmt.Sprintf("uefivars.py -i json -o edk2 -O %s", efiVarsPath)
+	}
+
+	// If "capsh" tool is present in the system, let's drop
+	// all capabilities except "cap_dac_override".
+	var cmdName string
+	var cmdArgs []string
+	if _, err := exec.LookPath("capsh"); err == nil {
+		cmdName = "capsh"
+		cmdArgs = []string{
+			"--drop=cap_chown,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read,cap_perfmon,cap_bpf,cap_checkpoint_restore",
+			"--", "-c", pyUEFIVarsCmd}
+	} else {
+		cmdName = "sh"
+		cmdArgs = []string{"-c", pyUEFIVarsCmd}
+	}
+
+	err := shared.RunCommandWithFds(context.TODO(), stdin, stdout, cmdName, cmdArgs...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// EFIVarsRead does not make any sense for containers.
+func (d *qemu) EFIVarsRead() (api.InstanceEFI, error) {
```

Comment:
We tend to return pointers to structs e.g. *api.InstanceEFI,this way we can return nil on error.

Reply:
Have done. Thanks!